text/0000-layer-table.md

```diff
@@ -0,0 +1,86 @@
+# Meta
+[meta]: #meta
+- Name: Move layer types to new table in layer.toml
+- Start Date: 2021-01-05
+- Author(s): natalieparellano
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Introduces a new table in `<layers>/<layer>.toml` to enable Bash buildpacks to easily opt in to layer re-use.
+
+# Definitions
+[definitions]: #definitions
+
+# Motivation
+[motivation]: #motivation
+
+[RFC 0052](https://github.com/buildpacks/rfcs/blob/main/text/0052-opt-in-layer-caching.md) described a change (to be implemented in Buildpack API 0.6) whereby buildpacks must opt in to layer re-use - i.e., each `<layers>/<layer>.toml` will be restored with no `launch`, `build`, or `cache` flags set and buildpacks must modify the restored TOML to set these flags.
+
+The RFC assumed that Bash buildpacks would be able to easily opt in to layer re-use by appending lines to the TOML, e.g.:
+
+`echo "launch = true" >> layer.toml`
+
+However, because `launch`, `build`, and `cache` are top level keys, the above command would add a `launch` key to the `[metadata]` table instead of the top level.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes moving the `launch`, `build`, and `cache` keys into a new `types` table in `<layers>/<layer>.toml`:
+
+```toml
+[types]
+launch = true
+build = false
+cache = false
+```
+
+A Bash buildpack could then write something like the following:
+
+```bash
+echo "[types]" >> layer.toml
+echo "launch = true" >> layer.toml
+```
+
+This change should be implemented with Buildpack API 0.6 so that it goes hand-in-hand with the opt in layer re-use feature.
+
+- If a buildpack implementing Buildpack API less than 0.6 tries to write `<layers>/<layer>.toml` with a `[types]` table, the lifecycle should fail.
+- If a buildpack implementing Buildpack API 0.6 or above tries to write `<layers>/<layer>.toml` with a top level `launch`, `build`, or `cache` key, the lifecycle should fail.
```

Comment:
Actually I'm not sure if this is necessary because the lifecycle won't ever present the file as it was written to a later build's buildpack... maybe a warning would be sufficient.

Reply:
Failure seems better to me, so buildpack authors get early explicit feedback that their top-level keys won't work with the new API.

Reply:
How would it work if you have a group of buildpacks where there are some that are `< 0.6` and `0.5`?

Reply:
> How would it work if you have a group of buildpacks where there are some that are < 0.6 and 0.5?

We can validate the layers from each buildpack using that buildpack's API. A group containing multiple APIs doesn't make this any more complicated.

Reply:
I want to say that the lifecycle would be responsible doing the conversion here. When restoring `<layers>/<layer>.toml` it would write the file in the format that each buildpack expects to find it - i.e., for older buildpacks, preserving the top-level keys, and for newer buildpacks, stripping them out. And the lifecycle when reading `<layers>/<layer>.toml` would parse it with the schema that is valid for each buildpack.